// @cerbos/api/lib/protobuf/cerbos/effect/v1/effect_pb.js
// Copyright 2021-2026 Zenauth Ltd.
// SPDX-License-Identifier: Apache-2.0
import { enumDesc, fileDesc } from "@bufbuild/protobuf/codegenv2";
/**
 * Describes the file cerbos/effect/v1/effect.proto.
 */
export const file_cerbos_effect_v1_effect = 
/*@__PURE__*/
fileDesc("Ch1jZXJib3MvZWZmZWN0L3YxL2VmZmVjdC5wcm90bxIQY2VyYm9zLmVmZmVjdC52MSpYCgZFZmZlY3QSFgoSRUZGRUNUX1VOU1BFQ0lGSUVEEAASEAoMRUZGRUNUX0FMTE9XEAESDwoLRUZGRUNUX0RFTlkQAhITCg9FRkZFQ1RfTk9fTUFUQ0gQA0JvChhkZXYuY2VyYm9zLmFwaS52MS5lZmZlY3RaPGdpdGh1Yi5jb20vY2VyYm9zL2NlcmJvcy9hcGkvZ2VucGIvY2VyYm9zL2VmZmVjdC92MTtlZmZlY3R2MaoCFENlcmJvcy5BcGkuVjEuRWZmZWN0YgZwcm90bzM");
/**
 * @generated from enum cerbos.effect.v1.Effect
 */
export var Effect;
(function (Effect) {
    /**
     * @generated from enum value: EFFECT_UNSPECIFIED = 0;
     */
    Effect[Effect["UNSPECIFIED"] = 0] = "UNSPECIFIED";
    /**
     * @generated from enum value: EFFECT_ALLOW = 1;
     */
    Effect[Effect["ALLOW"] = 1] = "ALLOW";
    /**
     * @generated from enum value: EFFECT_DENY = 2;
     */
    Effect[Effect["DENY"] = 2] = "DENY";
    /**
     * @generated from enum value: EFFECT_NO_MATCH = 3;
     */
    Effect[Effect["NO_MATCH"] = 3] = "NO_MATCH";
})(Effect || (Effect = {}));
/**
 * Describes the enum cerbos.effect.v1.Effect.
 */
export const EffectSchema = 
/*@__PURE__*/
enumDesc(file_cerbos_effect_v1_effect, 0);
