// @cerbos/core/lib/types/external/PolicySource.js
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceBlob}.
 */
export function policySourceIsBlob(source) {
    return source.kind === "blob";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceDatabase}.
 */
export function policySourceIsDatabase(source) {
    return source.kind === "database";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceDisk}.
 */
export function policySourceIsDisk(source) {
    return source.kind === "disk";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceEmbeddedPDP}.
 */
export function policySourceIsEmbeddedPDP(source) {
    return source.kind === "embeddedPDP";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceGit}.
 */
export function policySourceIsGit(source) {
    return source.kind === "git";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHub}.
 */
export function policySourceIsHub(source) {
    return source.kind === "hub";
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubDeployment}.
 *
 * @deprecated Use {@link policySourceIsHubRemoteBundle} instead.
 */
export function policySourceIsHubDeployment(source) {
    return policySourceIsHub(source) && "deploymentId" in source;
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubEmbeddedBundle}.
 */
export function policySourceIsHubEmbeddedBundle(source) {
    return policySourceIsHub(source) && "embeddedBundle" in source;
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubLabel}.
 */
export function policySourceIsHubLabel(source) {
    return policySourceIsHub(source) && "label" in source;
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubLocalBundle}.
 */
export function policySourceIsHubLocalBundle(source) {
    return policySourceIsHub(source) && "localBundle" in source;
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubPlayground}.
 */
export function policySourceIsHubPlayground(source) {
    return policySourceIsHub(source) && "playgroundId" in source;
}
/**
 * Type guard to check if a {@link PolicySource} is a {@link PolicySourceHubRemoteBundle}.
 *
 * @remarks
 * Requires the Cerbos policy decision point server to be at least v0.52.
 */
export function policySourceIsHubRemoteBundle(source) {
    return policySourceIsHub(source) && "remoteBundle" in source;
}
