// @cerbos/core/lib/types/external/Policy.js
/**
 * Type guard to check if a {@link Policy} is a set of {@link DerivedRoles}.
 */
export function policyIsDerivedRoles(policy) {
    return "derivedRoles" in policy;
}
/**
 * Type guard to check if a {@link Policy} is a set of {@link ExportConstants}.
 */
export function policyIsExportConstants(policy) {
    return "exportConstants" in policy;
}
/**
 * Type guard to check if a {@link Policy} is a set of {@link ExportVariables}.
 */
export function policyIsExportVariables(policy) {
    return "exportVariables" in policy;
}
/**
 * Type guard to check if a {@link Policy} is a {@link PrincipalPolicy}.
 */
export function policyIsPrincipalPolicy(policy) {
    return "principalPolicy" in policy;
}
/**
 * Type guard to check if a {@link Policy} is a {@link ResourcePolicy}.
 */
export function policyIsResourcePolicy(policy) {
    return "resourcePolicy" in policy;
}
/**
 * Type guard to check if a {@link Policy} is a {@link RolePolicy}.
 */
export function policyIsRolePolicy(policy) {
    return "rolePolicy" in policy;
}
