// @cerbos/core/lib/types/external/Status.js
/**
 * Status codes returned by the Cerbos policy decision point server.
 */
export var Status;
(function (Status) {
    /**
     * The operation was aborted.
     */
    Status[Status["ABORTED"] = 10] = "ABORTED";
    /**
     * The entity that the client attempted to create already exists.
     */
    Status[Status["ALREADY_EXISTS"] = 6] = "ALREADY_EXISTS";
    /**
     * The operation was cancelled.
     */
    Status[Status["CANCELLED"] = 1] = "CANCELLED";
    /**
     * The operation resulted in unrecoverable data loss or corruption.
     */
    Status[Status["DATA_LOSS"] = 15] = "DATA_LOSS";
    /**
     * The operation timed out.
     */
    Status[Status["DEADLINE_EXCEEDED"] = 4] = "DEADLINE_EXCEEDED";
    /**
     * The operation was rejected because the system is not in a state required for the operation's execution.
     */
    Status[Status["FAILED_PRECONDITION"] = 9] = "FAILED_PRECONDITION";
    /**
     * The operation failed due to an internal error.
     */
    Status[Status["INTERNAL"] = 13] = "INTERNAL";
    /**
     * The operation was rejected because an argument was invalid.
     */
    Status[Status["INVALID_ARGUMENT"] = 3] = "INVALID_ARGUMENT";
    /**
     * The requested entity was not found.
     */
    Status[Status["NOT_FOUND"] = 5] = "NOT_FOUND";
    /**
     * The operation completed successfully.
     */
    Status[Status["OK"] = 0] = "OK";
    /**
     * The operation was attempted past the valid range.
     */
    Status[Status["OUT_OF_RANGE"] = 11] = "OUT_OF_RANGE";
    /**
     * The caller does not have permission to execute the specified operation.
     */
    Status[Status["PERMISSION_DENIED"] = 7] = "PERMISSION_DENIED";
    /**
     * The operation failed because a resource has been exhausted.
     */
    Status[Status["RESOURCE_EXHAUSTED"] = 8] = "RESOURCE_EXHAUSTED";
    /**
     * The operation was rejected because it did not have valid authentication credentials.
     */
    Status[Status["UNAUTHENTICATED"] = 16] = "UNAUTHENTICATED";
    /**
     * The operation failed because the service is unavailable.
     */
    Status[Status["UNAVAILABLE"] = 14] = "UNAVAILABLE";
    /**
     * The operation is not supported.
     */
    Status[Status["UNIMPLEMENTED"] = 12] = "UNIMPLEMENTED";
    /**
     * An unknown error occurred.
     */
    Status[Status["UNKNOWN"] = 2] = "UNKNOWN";
})(Status || (Status = {}));
